src/mgs/mgs_fs_parser.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2018 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use crate::{
+    base_parsers::{period, target, word},
+    types::{FsName, Param, Record, Target, TargetStat, TargetStats, TargetVariant},
+};
+use combine::{
+    attempt,
+    error::ParseError,
+    many1,
+    parser::char::{newline, string},
+    stream::Stream,
+    Parser,
+};
+use std::collections::HashMap;
+
+pub fn params() -> Vec<String> {
+    vec!["mgs.*.live.*".to_string()]
+}
+
+#[derive(Debug)]
+enum MgsFsStat {
+    MgsFsNames(Vec<(Target, FsName)>),
+}
+
+fn fsname<I>() -> impl Parser<I, Output = (Target, FsName)>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    (
+        attempt(string("mgs")).skip(period()),
+        target().skip(period()),
+        string("live").skip(period()),
+        word().skip(newline()).map(FsName),
+    )
+        .map(|(_, target, _, fsname)| (target, fsname))
+}
+
+fn fsnames<I>() -> impl Parser<I, Output = Vec<(Target, FsName)>>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    many1(fsname())
+}
+
+fn mgs_fs_stat<I>() -> impl Parser<I, Output = (Param, MgsFsStat)>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    fsnames()
+        .map(|xs| {
+            let xs: Vec<(Target, FsName)> = xs
+                .into_iter()
+                .filter(|(_, name)| name.0 != "params")
+                .collect();
+
+            xs
+        })
+        .map(|xs| (Param("fsnames".into()), MgsFsStat::MgsFsNames(xs)))
+}
+
+pub fn parse<I>() -> impl Parser<I, Output = Vec<Record>>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    (mgs_fs_stat())
+        .map(|(param, stat)| match stat {
+            MgsFsStat::MgsFsNames(xs) => {
+                let mgs_map: HashMap<Target, Vec<FsName>> =
+                    xs.into_iter()
+                        .fold(HashMap::new(), |mut acc, (target, fs_name)| {
+                            if let Some(fs_names) = acc.get(&target) {
```

Comment:
nit, this would probably be cleaner by using the entry api.

Reply:
cool thanks for the tip. I updated to use this instead.